Update TTS calculation method

ChanWooY-1QBit opened the pull request:
Please see the [contributing guidelines](https://github.com/ccvm/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

> This PR is to update the TTS calculation method. The new method does not use arbitrary values of `best_energy` and is faster.
> `def generate_plot_data()` is now renamed with `def generated_TTS_plot_data()`
> `TTS_type` is not an attribute of the `ProblemMetadata` anymore, and now it is one of the inputs of `def generate_TTS_plot_data()`. 
> Related testcases are updated


### References

> #128 

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why.
>
> Please include details of the environment this PR was developed in.

- [√ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [√ ] I have added documentation for new/changed functionality in this PR
- [ √] All active GitHub checks for tests, formatting, and security are passing
- [ √] The correct base branch is being used, if not `main`

FarhadK-1QBit commented on ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py:
```diff
@@ -83,69 +81,26 @@ def ingest_metadata(self, metadata_filepath: str) -> None:
-        the optimal type and problem size inputs.
+    def generate_TTS_plot_data(
+        self,
+        method="advanced",
```
The variable `method` is not used and it's not really needed. In all the calculations of TTS we should just use the "advanced" method. We can remove this variable.

FarhadK-1QBit commented on ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py:
```diff
@@ -159,32 +114,73 @@ def generate_plot_data(self) -> pd.DataFrame:
+                    elif TTSType(TTS_type) == TTSType.physical:
+                        pp_time = np.mean(matching_df["pp_time"].values)
+                        iterations = matching_df["iterations"].values
+                        roundtrip_time = (
+                            (
+                                device_parameters["FPGA_fixed"]
+                                + device_parameters["FPGA_var_fac"]
+                                * float(problem_size)
+                            )
+                            * device_parameters["FPGA_clock"]
+                            + float(problem_size) * device_parameters["laser_clock"]
+                            + device_parameters["buffer_time"]
```
This calculation of `roundtrip_time` is specific to the MF-CCVM solver. For the DL-CCVM solver, it should be written as:
```
roundtrip_time = float(problem_size) * device_parameters["laser_clock"]
```

ChanWooY-1QBit replied:
Does this mean that when we generate TTS data for plotting, the plotter needs to identify which solver was used because the `rounttrip_time` calculation depends on it?

ChanWooY-1QBit replied:
This `machine_time` calculation is replaced with a callback function. Thus, the `roundtrip_time` will be moved into the callback function.

FarhadK-1QBit reviewed (commented):
Thank you for these changes, Chan Woo! 

I'm wondering if we need to make some changes to the calculation of TTS (which would also affect the ETS calculation).
As we discussed, maybe we can remove the variable `TTS_type` and replace it with a variable named `machine` or `device` which, when it's equal to `dl-ccvm` or `mf-ccvm` it would calculate the TTS according to the `roundtrip_time` calculation I mentioned in my comment in this review. 
I will post a detailed review of this in the ccvm-boxqp channel.

KatieO4t1qbit commented on examples/ccvm_boxqp_plot.py:
```diff
@@ -50,10 +50,17 @@
         metadata_filepath=metadata_filepath,
         problem="BoxQP",
-        TTS_type="wallclock",
+        machine_time_func=solver.cpu_machine_time(
```
I believe this won't run until we have the solver bits written, too, and we don't want to merge something that can't be run. However, I like that this PR is staying within scope, thank you.
My recommendation to merge this fast is to add an example of how the user can define their own cpu_machine_time() function _in this file_. As long as this file runs, the exact implementation of that function doesn't matter, but for the sake of making the best code we can let's use the cpu implementation.

FarhadK-1QBit reviewed (approved):
These look good to me, Chan Woo! Thank you!

KatieO4t1qbit commented on ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
         """Test BoxQP Metadata class "ingest_result_data" method when valid
         inputs are given.
 
         # TODO: Implementation
```
rm this comment now that the implementation is in :)

KatieO4t1qbit commented on ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
         """Test BoxQP Metadata class "ingest_result_data" method when invalid
         inputs are given.
 
         # TODO: Implementation
```
same here! And there are two more instances below that can go.

KatieO4t1qbit commented on ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
+        boxqp_metadata = BoxQPMetadata(self.valid_problem_type)
+        boxqp_metadata.ingest_metadata(self.valid_metadata_filepath)
+
+        self.assertGreater(boxqp_metadata._BoxQPMetadata__batch_size, 0)
+        self.assertGreater(len(boxqp_metadata._BoxQPMetadata__problem_size_list), 0)
```
Will these values be the same every time, since we are using some set metadata? If so, let's check the exact value.

KatieO4t1qbit reviewed (approved):
Great work, Chan Woo!
I left a few minor suggestions that won't affect the code itself, so I'll approve it but let me know if anything else changes and you'd like me to take another look :)

ChanWooY-1QBit commented:
@KatieO4t1qbit I addressed all your comments, but if you can take a quick look on the `test_BoxQP_metadata.py` file `example/ccvmp_boxqp_plot.py` files, then it would be great. Thanks!

KatieO4t1qbit commented on examples/ccvm_boxqp_plot.py:
```diff
@@ -50,20 +52,55 @@
-        os.makedirs(PLOT_OUTPUT_DIR)
-        print("Plot folder doesn't exist yet. Creating: ", PLOT_OUTPUT_DIR)
+    # Save to local
+    tts_plot_fig.savefig(PLOT_OUTPUT_DEST)
```
This seems to be the figure that got saved in the TTS plot file, but it looks like ETS to me. Can you check if this happens on your side?

![DL-CCVM_TTS_cpu_plot](https://github.com/1QB-Information-Technologies/ccvm/assets/19154892/ef61a459-f0d2-4055-87f0-3ca2c8058b5e)

ChanWooY-1QBit replied:
I am getting the same plots.
![ets](https://github.com/1QB-Information-Technologies/ccvm/assets/77126963/8e4049f3-9ad2-482b-a7f8-e6c78d000f8a)
![tts](https://github.com/1QB-Information-Technologies/ccvm/assets/77126963/e734b30f-1ed9-436e-9ada-f4b3175f446c)

KatieO4t1qbit replied:
Hmm, okay let me try again!

KatieO4t1qbit replied:
Okay I think I found the issue: both plots get saved to the same place: PLOT_OUTPUT_DEST (which we don't modify):
```
ets_plot_fig.savefig(PLOT_OUTPUT_DEST)
```

I think we need a variable here or a second constant.

ChanWooY-1QBit replied:
I separated the ETS and TTS plot output destination. And here are the new plots:
![ets](https://github.com/1QB-Information-Technologies/ccvm/assets/77126963/dfa87550-00dd-4f96-a99f-b7545e299148)
![tts](https://github.com/1QB-Information-Technologies/ccvm/assets/77126963/a7c64e78-7cd8-497c-bc9a-4afeed8af713)

ChanWooY-1QBit commented on ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py:
```diff
@@ -35,52 +64,78 @@ def test_BoxQP_metadata_invalid_problem_type(self):
 
-    def test_generate_arrays_of_TTS_method_invalid(self):
-        """Test BoxQP Metadata class "generate_arrays_of_TTS" method when
+        ETS_plot_data = boxqp_metadata.generate_plot_data(
```
It really depends on what metric function you provide. The same `generate_plot_data` method is used for both ETS and TTS plot data calculation, but the output depends on whether it is the machine time calc function or the energy max calc function.

KatieO4t1qbit replied:
Oh I see. I had a misunderstanding. I didn't realize that the function spits out the final value of the metric that we will be plotting, I thought it spit out a value that was used as _part_ of the metrics calculation.
Side note: with my new understanding, the name `metric_func` makes far more sense.

I think we might still want to improve clarity for the user, i.e. what if they specify an energy function here but then call TTS formatting? Given our current timelines, I'm happy for that to be addressed in a future PR, and we can ignore this comment for now.

KatieO4t1qbit replied:
Made a ticket! We can revisit later :)
https://github.com/1QB-Information-Technologies/ccvm/issues/163

KatieO4t1qbit reviewed (approved):
Awesome, looks good Chan Woo! 🔥
